core/drivers/crypto/hisilicon/hpre_dh.c

```diff
@@ -0,0 +1,379 @@
+// SPDX-License-Identifier: BSD-2-Clause
+/*
+ * Copyright (c) 2022, HiSilicon Technologies Co., Ltd.
+ * Kunpeng hardware accelerator hpre dh algorithm implementation.
+ */
+
+#include "hpre_main.h"
+#include "hpre_dh.h"
+
+static TEE_Result hpre_dh_alloc_keypair(struct dh_keypair *key,
+					size_t size_bits)
+{
+	if (!key || !size_bits) {
+		EMSG("Invalid input parameter!");
+		return TEE_ERROR_BAD_PARAMETERS;
+	}
+
+	memset(key, 0, sizeof(*key));
+	key->g = crypto_bignum_allocate(size_bits);
+	if (!key->g)
+		goto g_err;
+
+	key->p = crypto_bignum_allocate(size_bits);
+	if (!key->p)
+		goto p_err;
+
+	key->x = crypto_bignum_allocate(size_bits);
+	if (!key->x)
+		goto x_err;
+
+	key->y = crypto_bignum_allocate(size_bits);
+	if (!key->y)
+		goto y_err;
+
+	/* Allocate subprime even if not used */
+	key->q = crypto_bignum_allocate(size_bits);
+	if (!key->q)
+		goto q_err;
+
+	return TEE_SUCCESS;
+q_err:
+	crypto_bignum_free(&key->y);
+y_err:
+	crypto_bignum_free(&key->x);
+x_err:
+	crypto_bignum_free(&key->p);
+p_err:
+	crypto_bignum_free(&key->g);
+g_err:
+	EMSG("HPRE dh alloc key pair fail.");
+
+	return TEE_ERROR_OUT_OF_MEMORY;
+}
+
+static enum hisi_drv_status hpre_dh_fill_sqe(void *bd, void *info)
+{
+	struct hpre_dh_msg *msg = (struct hpre_dh_msg *)info;
+	struct hpre_sqe *sqe = (struct hpre_sqe *)bd;
+
+	sqe->w0 = msg->alg_type | (0x1 << HPRE_DONE_SHIFT);
+	sqe->task_len1 = TASK_LENGTH(msg->key_bytes);
+	sqe->key = msg->x_p_dma;
+	sqe->in = msg->g_dma;
+	sqe->out = msg->out_dma;
+
+	return HISI_QM_DRVCRYPT_NO_ERR;
+}
+
+static enum hisi_drv_status hpre_dh_parse_sqe(void *bd, void *info)
+{
+	struct hpre_dh_msg *msg = (struct hpre_dh_msg *)info;
+	struct hpre_sqe *sqe = (struct hpre_sqe *)bd;
+	uint16_t err, done;
+
+	err = HPRE_TASK_ETYPE(sqe->w0);
+	done = HPRE_TASK_DONE(sqe->w0);
+	if (done != HPRE_HW_TASK_DONE || err) {
+		EMSG("HPRE do dh fail! done=0x%x, etype=0x%x", done, err);
+		return HISI_QM_DRVCRYPT_IN_EPARA;
+	}
+
+	if (hpre_bin_to_crypto_bin(msg->out, msg->out, msg->key_bytes,
+				   msg->out_bytes)) {
+		EMSG("Fail to transfer dh_y from hpre_bin to crypto_bin");
+		return HISI_QM_DRVCRYPT_EINVAL;
+	}
+
+	return HISI_QM_DRVCRYPT_NO_ERR;
+}
+
+static TEE_Result hpre_dh_do_task(void *msg)
+{
+	struct hisi_qp *dh_qp;
+	TEE_Result res = 0;
+	int32_t ret;
+
+	dh_qp = hpre_create_qp(HISI_QM_CHANNEL_TYPE0);
+	if (!dh_qp) {
+		EMSG("Fail to create dh qp!");
+		return TEE_ERROR_BUSY;
+	}
+
+	dh_qp->fill_sqe = hpre_dh_fill_sqe;
```

Comment:
For how long are these pointers supposed to be valid, that is, when are these fields re-initialized?

Reply:
These fields will be re-initialized when this qp is used for other algorithm (after release_qp)